core/src/main/java/org/apache/brooklyn/policy/InvokeEffectorOnSetChange.java

```diff
@@ -0,0 +1,203 @@
+
+/**
+ * Subscribes to events on a collection {@link AttributeSensor} and invokes the named
+ * effectors for each element that was added and removed.
+ * <p>
+ * The policy only detects <em>replacements</em> of the collection; it does not act on
+ * modifications. If the sensor has value <code>A</code> and an element is added &ndash;
+ * value <code>A'</code> &ndash; the on-added effector is not invoked. If the sensor is
+ * later set to <code>B</code> the delta is made between <code>A</code> and <code>B</code>,
+ * not <code>A'</code> and <code>B</code>.
+ * <p>
+ * To simplify the detection of additions and removals the collection is converted to a
+ * {@link Set}. This means that only a single event will fire for duplicate elements in
+ * the collection.
+ * <p>
+ * The effectors are provided the elements that changed in their parameter map. If the
+ * sensor is a collection of maps the elements are provided with their keys coerced to
+ * strings and their values unchanged. Otherwise the elements are provided in a
+ * single-entry map keyed by the value for {@link #PARAMETER_NAME}.
+ */
+public class InvokeEffectorOnSetChange extends AbstractPolicy implements SensorEventListener<Collection<?>> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(InvokeEffectorOnSetChange.class);
+
+    public static final ConfigKey<AttributeSensor<? extends Collection<?>>> TRIGGER_SENSOR = ConfigKeys.newConfigKey(
+            new TypeToken<AttributeSensor<? extends Collection<?>>>() {},
+            "sensor",
+            "Sensor to be monitored.");
+
+    public static final ConfigKey<String> PREVIOUS_SENSOR_NAME = ConfigKeys.newStringConfigKey(
+            "previousSensorName",
+            "The name under which the previous value for the trigger sensor should be published. " +
+                    "If unset \".previous\" will be appended to the name of the trigger sensor.");
+
+    public static final ConfigKey<String> ON_ADDED_EFFECTOR_NAME = ConfigKeys.newStringConfigKey(
+            "onAdded",
+            "Name of the effector to invoke when entries are added to the collection.");
+
+    public static final ConfigKey<String> ON_REMOVED_EFFECTOR_NAME = ConfigKeys.newStringConfigKey(
+            "onRemoved",
+            "Name of the effector to invoke when entries are removed from the collection.");
+
+    public static final ConfigKey<String> PARAMETER_NAME = ConfigKeys.newStringConfigKey(
+            "parameterName",
+            "The name of the parameter to supply to the effectors",
+            "value");
+
+    /** The previous version of the set against which events will be compared. */
+    private Set<Object> previous = Collections.emptySet();
+
+    /** Guards accesses of previous. */
+    private final Object[] updateLock = new Object[0];
+
+    @Override
+    public void setEntity(EntityLocal entity) {
+        super.setEntity(entity);
+        Sensor<? extends Collection<?>> sensor =
```

Comment:
Worth a comment that the method will be executed on rebind. In case of adding additional validation should be backwards compatible with existing persistence state.